File: py/147-insertion-sort-list.py
"""
 Source : https://oj.leetcode.com/problems/insertion-sort-list/
 Author : Alex Xu
 Date   : Apr 30, 2015
 Problem: Sort a linked list using insertion sort.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return None
        
        dummy = ListNode(0)
        dummy.next, head = head, head.next
        dummy.next.next = None
        
        while head:
            head = self.insert_to_sorted_list(dummy, head)
            
        return dummy.next
    
    def insert_to_sorted_list(self, dummy, node):
        current = dummy
        
        while current.next and current.next.val < node.val:
            current = current.next
        
        current.next, node.next, node = node, current.next, node.next
        
        return node

File: java/147-insertion-sort-list.java
/*
 Source : https://oj.leetcode.com/problems/insertion-sort-list/
 Author : Alex Xu
 Date   : May 01, 2018
 Problem: Sort a linked list using insertion sort.
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        // using a dummy node makes insertion much easier
        ListNode dummy = new ListNode(0);
        // head.next, the next node in the original list to be processed
        ListNode next = null;
        
        while(head != null) {
            // save head.next since head will be relinked during insertion
            next = head.next;
            dummy.next = insertNodeToSortList(dummy, head);
            head = next;
        }
        
        return dummy.next;
    }
    
    public ListNode insertNodeToSortList(ListNode dummy, ListNode node) {
        ListNode current = dummy;
        
        while (current.next != null && current.next.val < node.val) {
            current = current.next;
        }
        
        node.next = current.next;
        current.next = node;

        return dummy.next;
    }
}


class Solution2 {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        // the list starting with head.next is now sorted
        // this is not in the order as descriped in the problem
        // but it resloves the problem.
        head.next = insertionSortList(head.next);
        
        // move head to the proper position
         return moveHeadNode(head);
    }
    
    public ListNode moveHeadNode(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        if (head.val < head.next.val) {
            return head;
        }
        
        ListNode next = head.next;
        head.next = next.next;
        next.next = moveHeadNode(head);
        
        return next;
    }
}
